### PR TITLE
RO opens pack before RW

### PR DESCRIPTION
**Pull request by icristescu**
Type raw in the Io module is now of type option raw. When a RO instance is opened first and there are no files on disks, then its raw field is set to `None`. RO will try to open the file only the first time it needs to read it.

RO instances do not refill their memory caches for Dict and the branch store when opened, but instead only when the first read occurs. To ensure this, the offset is set to `0` when a file is opened.

### CODE REVIEW

File: src/irmin-pack/IO.ml [hunk 1]
```diff
@@ -232,25 +259,27 @@ module Unix : S = struct
...
-        Raw.unsafe_set_offset raw 0L;
-        Raw.unsafe_set_version raw current_version;
-        v ~offset:0L ~version:current_version raw
+        if readonly then v ~offset:0L ~version:current_version None
+        else
+          let x = Unix.openfile file Unix.[ O_CREAT; mode; O_CLOEXEC ] 0o644 in
+          let raw = Raw.v x in
+          Raw.unsafe_set_offset raw 0L;
+          Raw.unsafe_set_version raw current_version;
+          v ~offset:0L ~version:current_version (Some raw)
```
Comment (CraigFe) [inline src/irmin-pack/IO.ml]: Totally subjective, but I'd prefer pushing the `if` further into the logic:

```ocaml
let raw_opt =
  match readonly with
  | true -> None
  | false ->
      let x = Unix.openfile file Unix.[ O_CREAT; mode; O_CLOEXEC ] 0o644 in
      let raw = Raw.v x in
      Raw.unsafe_set_offset raw 0L;
      Raw.unsafe_set_version raw current_version;
      Some raw
in
v ~offset:0L ~version:current_version raw_opt

Comment (CraigFe) [reply src/irmin-pack/IO.ml]: There's also a small TOCTOU race condition here (which was already present): the file may be created and filled with data after line 260 but before line 264.

I think we should pass the `O_EXCL` flag and drop into the `true` case if it raises a "File already exists" exception.

File: src/irmin-pack/IO.ml [hunk 2]
```diff
@@ -232,25 +259,27 @@ module Unix : S = struct
...
         let x = Unix.openfile file Unix.[ O_EXCL; mode; O_CLOEXEC ] 0o644 in
         let raw = Raw.v x in
         if fresh then (
           Raw.unsafe_set_offset raw 0L;
           Raw.unsafe_set_version raw current_version;
-          v ~offset:0L ~version:current_version raw )
+          v ~offset:0L ~version:current_version (Some raw) )
         else
-          let offset = Raw.unsafe_get_offset raw in
+          let offset = if readonly then 0L else Raw.unsafe_get_offset raw in
```
Comment (CraigFe) [inline src/irmin-pack/IO.ml]: This change isn't clear to me; why the different behaviour for RO instances if we know that `raw` exists? (is it just an unrelated optimisation?)

File: src/irmin-pack/IO.ml [hunk 3]
```diff
@@ -232,25 +259,27 @@ module Unix : S = struct
...
-        v ~offset:0L ~version:current_version raw
+        if readonly then v ~offset:0L ~version:current_version None
+        else
+          let x = Unix.openfile file Unix.[ O_CREAT; mode; O_CLOEXEC ] 0o644 in
+          let raw = Raw.v x in
+          Raw.unsafe_set_offset raw 0L;
+          Raw.unsafe_set_version raw current_version;
+          v ~offset:0L ~version:current_version (Some raw)
     | true ->
         let x = Unix.openfile file Unix.[ O_EXCL; mode; O_CLOEXEC ] 0o644 in
```
Comment (CraigFe) [inline src/irmin-pack/IO.ml]: As mentioned on Slack, I think this `O_EXCL` may be an existing bug.

File: src/irmin-pack/IO.ml [hunk 4]
```diff
@@ -117,28 +117,50 @@ module Unix : S = struct
...
+
+  (* only RO instances could not have the file opened yet. *)
+  let try_load_file t =
+    match t.raw with
+    | Some _ -> ()
+    | None ->
+        Log.debug (fun l -> l "checking on-disk %s" t.file);
+        if Sys.file_exists t.file then (
+          let x =
+            Unix.openfile t.file Unix.[ O_EXCL; O_RDONLY; O_CLOEXEC ] 0o644
```
Comment (CraigFe) [inline src/irmin-pack/IO.ml]:
```suggestion
            Unix.openfile t.file Unix.[ O_RDONLY; O_CLOEXEC ] 0o644
```
Not sure what the `O_EXCL` flag is achieving here.

File: src/irmin-pack/IO.ml [hunk 5]
```diff
@@ -117,28 +117,50 @@ module Unix : S = struct
 
   type t = {
     file : string;
-    mutable raw : Raw.t;
+    mutable raw : Raw.t option;
```
Comment (CraigFe) [inline src/irmin-pack/IO.ml]: Perhaps we should add a comment to explain the semantics of this option? Something like "may be `None` if no read-write instances have been constructed".

(I thought about suggesting a type that only permits a `None` → `Some x` transition, but the comment probably suffices)

File: src/irmin-pack/IO.ml [hunk 6]
```diff
@@ -117,28 +117,50 @@ module Unix : S = struct
...
+    | Some _ -> ()
+    | None ->
+        Log.debug (fun l -> l "checking on-disk %s" t.file);
+        if Sys.file_exists t.file then (
+          let x =
+            Unix.openfile t.file Unix.[ O_EXCL; O_RDONLY; O_CLOEXEC ] 0o644
+          in
+          let raw = Raw.v x in
+          let version = Raw.unsafe_get_version raw in
+          assert (version = t.version);
```
Comment (CraigFe) [inline src/irmin-pack/IO.ml]: I assume this was taken from elsewhere, but the use of `assert` is suspicious. Isn't this visible to the user if the version is changed concurrently?

File: src/irmin-pack/IO.ml (hunk 1 above)
Comment (icristescu) [reply src/irmin-pack/IO.ml]: > There's also a small TOCTOU race condition here (which was already present): the file may be created and filled with data after line 260 but before line 264.

Good catch. I think the race can only occur when two rw instances (of the same process) try to open a "fresh" pack concurrently. In this case, it should be ok, they write the same thing.
That's the only case I see, because as soon as one rw instance opened a pack, it is then cached for future instances.

File: src/irmin-pack/IO.ml (hunk 2 above)
Comment (icristescu) [reply src/irmin-pack/IO.ml]: It's related to https://github.com/mirage/irmin/pull/955/files#diff-1beb7e2142825f197755316a39d00699R227 : before, when an ro opened an existing file, it refilled its buffers with what was in that file starting from offset `0`. 
Now, an ro cannot do a refill when opening a file (as it might not exists) and so it needs to do it when it first reads a value. At that point, it compares its offset with the real offset (https://github.com/mirage/irmin/pull/955/files#diff-1beb7e2142825f197755316a39d00699L165). To ensure that it does sync with the on-disk file, we have to set the offset to `0` when opening the file in ro. Does it make sense?  

Agree that it is not clear, I should probably add a comment in the code about this.

File: src/irmin-pack/IO.ml (hunk 6 above)
Comment (icristescu) [reply src/irmin-pack/IO.ml]: I think that the point of having a `version`, is to ensure that only files compatible with the current version of pack are read. It should not be possible to change it "on-the-fly".

File: src/irmin-pack/IO.ml (hunk 1 above)
Comment (CraigFe) [reply src/irmin-pack/IO.ml]: Do they always write the same thing? The two RW instances might pass different `~version`s.

Comment (icristescu) [reply src/irmin-pack/IO.ml]: `version` is the `current_version` which is hardcoded in `pack.ml`. And I think it will stay this way but I'm not really sure, @pascutto, @samoht ?

File: src/irmin-pack/IO.ml (hunk 6 above)
Comment (CraigFe) [reply src/irmin-pack/IO.ml]: If a user upgrades `Irmin_pack` and tries to read an old file, will they see an `assert` failure here? My thinking was that we might want a `Corrupted` or `Unsupported_version` exception here instead, rather than putting such checks in `asserts` where (a) they can be compiled away and (b) they look like constraints that have already been ensured by the code (which I don't think is the case here).

File: src/irmin-pack/IO.ml (hunk 1 above)
Comment (samoht) [reply src/irmin-pack/IO.ml]: It can change if we decide to change the format of irmin-pack.

File: src/irmin-pack/IO.ml (hunk 6 above)
Comment (icristescu) [reply src/irmin-pack/IO.ml]: Ah, ok, I get your point now (the concurrent part confused me, sorry). A more informative failure is definitely better, I'll add it!

File: src/irmin-pack/IO.ml (hunk 1 above)
Comment (samoht) [reply src/irmin-pack/IO.ml]: But we haven't implemented yet the logics to handle such cases : we need to do something if the version on disk is different than the version we expect ; a simple solution will just be to fail with an informative error message -- a more complex solution would be to upgrade the storage to the new on-disk format.

Comment (CraigFe) [reply src/irmin-pack/IO.ml]: In that case, they will always pass the same thing :slightly_smiling_face: If we require all constructed instances to have the same version, I suggest that we do one of:
 - detail this constraint in a doc-comment of `IO.S.v` (and comment here that the constraint is being relied upon)
 - move `Irmin_pack.current_version` to a central `Version` module avoid passing it in at all

Comment (samoht) [reply src/irmin-pack/IO.ml]: Not if you run 2 different instance of irmin-pack concurrently (with different on-disk format)

Comment (CraigFe) [reply src/irmin-pack/IO.ml]: Good point; I forgot this was a cross-process race too. In that case I revert to my earlier suggestion of a defensive `O_EXCL` flag.

File: src/irmin-pack/IO.ml (hunk 2 above)
Comment (CraigFe) [reply src/irmin-pack/IO.ml]: I think I understand, thanks :) However, I think in this particular case we _know_ that the file exists (since we've just opened it), right?

Comment (icristescu) [reply src/irmin-pack/IO.ml]: Yes, but if you want to check that the file exists from `irmin-pack.ml` you need to add a `if Sys.file_exists` in the code here https://github.com/mirage/irmin/pull/955/files#diff-1beb7e2142825f197755316a39d00699R227 (which I don't want to do, I think it breaks a bit the abstraction of file). 
By having the offset set to `0` for all ro instances when they are created, we avoid checking this. We have to do `refill` anyway, it is just postponed until the first read.

Comment (CraigFe) [reply src/irmin-pack/IO.ml]: I'm not proposing changing it in `irmin-pack.ml`, just here where it's not strictly necessary.

As you point out, it doesn't really make much of a difference anyway. Let's not worry about it :slightly_smiling_face:

File: src/irmin-pack/IO.ml (hunk 3 above)
Comment (icristescu) [reply src/irmin-pack/IO.ml]: I've removed `O_EXCL` when the file exists and add it to when the file does not exists so that it can be used with `O_CREAT`. 
It seems that it is also a bug in [index_unix](https://github.com/mirage/index/blob/master/src/unix/index_unix.ml#L305)

**Comment by icristescu** [comment]
Thanks for your comments, I've rebased it. (I think it makes sense to review and merge it, before opening a new PR where the irmin-pack uses the index's IO module.)

**Comment by icristescu** [comment]
I've rebased this. Some extra changes (in addition to the ones mentioned above) introduced by this PR:

- the RO instances are not sync-ed with the disk after they are opened: an extra `sync` call is necessary;
- the RO instances have offset 0 when opening a file; to get the actual offset, we have to do an extra `force_offset`

**Comment by samoht** [comment]
This is still an issue ; @icristescu can you rebase it? @Ngoguey42 can you review it ?

File: src/irmin-pack/IO.ml (hunk 1 above)
Comment (icristescu) [reply src/irmin-pack/IO.ml]: I added the `O_EXCL` flag here (and removed it from when we [open the file and the file exists](https://github.com/mirage/irmin/pull/955/files#diff-1a4152c6731a5518a2728ab388cf55d2b58315685a58019c5f0eb8f30ce6b01dR245)).

**Comment by icristescu** [comment]
Rebased

**Comment by icristescu** [comment]
Rebased.

**Review by Ngoguey42 — APPROVED** [review]
LGTM

File: src/irmin-pack/IO.ml [hunk 7]
```diff
@@ -238,15 +269,22 @@ module Unix : S = struct
           match actual_version with
           | `V1 ->
               Log.debug (fun l -> l "[%s] file exists in V1" file);
-              let offset = Raw.Offset.get raw in
-              v ~offset ~version:`V1 ~generation:Int63.zero raw
+              let offset =
+                if readonly then Int63.zero else Raw.Offset.get raw
```
Comment (samoht) [inline src/irmin-pack/IO.ml:273]: this is a bit odd, why is this 0 in the read-only case?

Comment (samoht) [reply src/irmin-pack/IO.ml:273]: ha it seems to be documented bellow - but I am not sure to understand the difference of behaviour between read-only and read-write here.

File: src/irmin-pack/IO_intf.ml [hunk 8]
```diff
@@ -16,11 +16,10 @@
 
 open! Import
 
-type headers = { offset : int63; generation : int63 }
```
Comment (samoht) [inline src/irmin-pack/IO_intf.ml:19]: why is this needed (just out of curiosity)

File: src/irmin-pack/atomic_write.ml [hunk 9]
```diff
@@ -220,6 +222,7 @@ module Make (Current : Version.S) (K : Irmin.Type.S) (V : Irmin.Hash.S) = struct
 
   let list t =
     Log.debug (fun l -> l "[branches] list");
+    if IO.readonly t.block then sync_offset t;
```
Comment (samoht) [inline src/irmin-pack/atomic_write.ml:225]: Why is this needed? Is it because we don't have a `sync` for atomic-write stores? I'm not sure why this is related to the ordering in which we open the RO/RW instances?

**Comment by samoht** [comment]
It seems that this PR is doing 2 different things: allowing empty read-only instances against *and* adding (missing?) synchronisation points in various places of the code.

While the former is a good change, for the later we need to be extra careful as it might have big performance impact: the read-only instances should sync only when requested as the application code can try to minimise the  number of such calls.

**Comment by icristescu** [comment]
In order to allow the RO to open the store before RW I need first to ensure that:
- the RO instances are not sync-ed with the disk right after they are opened: one extra sync call is necessary after opening the store - we need this in the case where the files are not yet on disk when opening the store 
- RO reads the files on the first `sync` - I did this by always setting the offset to `0` when opening the store, even if the files are already on disk. But then there are some operations that require the "actual" offset (when calling directly IO in the `Checks` module and when reconstructing the index).  

Unrelated to the RO change, this PR also fixes the O_EXCL flag. I'll extract that in a separate PR and split the rest into 2 smaller PRs (I agree that we need to be extra careful with these changes :) )

**Comment by samoht** [comment]
> the RO instances are not sync-ed with the disk right after they are opened:

I think this is ok? It means that there is a race between the RO and RW open, and the RW won. The RO will then need an extra sync (coming from the application) to be up-to-date, and/or the application will have to make sure there is no race.

**Comment by Ngoguey42** [comment]
Closing because we don't need this and because the code changed drastically since.
